### PR TITLE
CMake build system

### PR DESCRIPTION
**Pull request by chriso**
This replaces the current `makefile` with a CMake build system.

There's a new `make install` target that installs headers and a shared library, which means that you can include the library using `LDFLAGS=-lroaring` and `#include <roaring/roaring.h>`.

All you need to do now is to run `cmake [OPTIONS] .` once to build the Makefile, followed by the usual `make [target]`. More CMake info here: http://llvm.org/releases/3.8.0/docs/CMake.html

For example, to install the library without AVX tuning, and to show the build commands:

```sh
$ cmake -DAVX_TUNING=0 .
$ make VERBOSE=1
$ make install
```

I've preserved everything as best I could. The tests and benchmarks are compiled into the top-level directory and can be invoked in exactly the same way, and the `run_unit.sh` script still works.

### CODE REVIEW

**Comment by lemire** [comment]
Cool stuff. 

@fsaintjacques : you want to take a shot at reviewing this PR? (at your leisure)

@chriso : please do not use this code for anything yet! It is not ready!

**Comment by chriso** [comment]
@lemire I'm not using the library yet. I was planning on finishing my own similar library (https://github.com/chriso/roaring-bitmap) but figured I'd just contribute some patches here.

**Comment by lemire** [comment]
@chriso  You are referenced here : http://roaringbitmap.org/ ;-)

**Comment by chriso** [comment]
Oh neat! :wink:

**Comment by lemire** [comment]
@chriso  I have been aware of your project for some time. I'd hate for you to drop it because of CRoaring. As I have argued with @fsaintjacques, I do hope that people will simply steal from CRoaring in the end. This being said, should you feel like joining forces, you look more than capable to help us... and I think I speak for the whole gang when I stay you'd be welcome. But, once more, I'd hate for you to drop your project because of us.

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+# invoke all tests with: make check
+add_custom_target(check COMMAND ${CMAKE_CTEST_COMMAND} --verbose DEPENDS
+    unit
+    bitset_container_unit
+    array_container_unit
+    mixed_container_unit
+    run_container_unit
+    toplevel_unit)
+
+add_executable (real_bitmaps_benchmark benchmarks/real_bitmaps_benchmark.c)
```
Comment (fsaintjacques) [inline CMakeLists.txt]: Does the `add_executable` directive affects the make install target, i.e. will benchmarks binaries be installed?

Comment (chriso) [inline CMakeLists.txt]: Not without an `install(PROGRAMS ...)` directive.

```sh
~repo/CRoaring chriso/cmake $ make install
[ 37%] Built target roaring
[ 43%] Built target array_container_benchmark
[ 50%] Built target array_container_unit
[ 56%] Built target bitset_container_benchmark
[ 62%] Built target bitset_container_unit
[ 68%] Built target mixed_container_unit
[ 75%] Built target real_bitmaps_benchmark
[ 81%] Built target run_container_benchmark
[ 87%] Built target run_container_unit
[ 93%] Built target toplevel_unit
[100%] Built target unit
Install the project...
-- Install configuration: ""
-- Installing: /usr/local/lib/libroaring.dylib
-- Installing: /usr/local/include/roaring/array_util.h
-- Installing: /usr/local/include/roaring/bitset_util.h
-- Installing: /usr/local/include/roaring/portability.h
-- Installing: /usr/local/include/roaring/roaring.h
-- Installing: /usr/local/include/roaring/roaring_array.h
-- Installing: /usr/local/include/roaring/utilasm.h
-- Installing: /usr/local/include/roaring/containers/array.h
-- Installing: /usr/local/include/roaring/containers/bitset.h
-- Installing: /usr/local/include/roaring/containers/containers.h
-- Installing: /usr/local/include/roaring/containers/convert.h
-- Installing: /usr/local/include/roaring/containers/mixed_intersection.h
-- Installing: /usr/local/include/roaring/containers/mixed_union.h
-- Installing: /usr/local/include/roaring/containers/run.h
```

Comment (chriso) [inline CMakeLists.txt]: I see your [`add_c_executable`](https://github.com/fsaintjacques/libtwiddle/blob/446c1ac28232436cd814386ab30f78bb9455c6e2/tools/cmake/FindCTargets.cmake#L149) function does call [`install`](https://github.com/fsaintjacques/libtwiddle/blob/446c1ac28232436cd814386ab30f78bb9455c6e2/tools/cmake/FindCTargets.cmake#L185).

The CMake build system from libtwiddle is far more sophisticated than the one I went with here. I tried to keep it simple and also maintain backwards compat with the current scripts. I'm happy to accept any suggestions or improvements :smile:

Comment (fsaintjacques) [inline CMakeLists.txt]: That's what I saw afterward. The credit is due to: https://github.com/redjack/c-template.

**Comment by fsaintjacques** [comment]
Seems good to me, one step closer to get travis going :)

**Comment by fsaintjacques** [comment]
ah wait, it doesn't compile on Linux.

**Comment by chriso** [comment]
@fsaintjacques what errors do you get?

I only have a basic Alpine VM setup at the moment. My build output:

```sh
~/CRoaring chriso/cmake $ uname -a
Linux alpine 4.1.18-0-grsec #1-Alpine SMP Wed Mar 2 12:21:10 GMT 2016 x86_64 Linux
~/CRoaring chriso/cmake $ cmake . && make check
-- The C compiler identification is GNU 5.3.0
-- The CXX compiler identification is GNU 5.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /root/CRoaring
Scanning dependencies of target roaring
[  4%] Building C object CMakeFiles/roaring.dir/src/array_util.c.o
[  8%] Building C object CMakeFiles/roaring.dir/src/bitset_util.c.o
[ 12%] Building C object CMakeFiles/roaring.dir/src/containers/array.c.o
[ 16%] Building C object CMakeFiles/roaring.dir/src/containers/bitset.c.o
/root/CRoaring/src/containers/bitset.c: In function 'bitset_container_create':
/root/CRoaring/src/containers/bitset.c:35:9: warning: implicit declaration of function 'posix_memalign' [-Wimplicit-function-declaration]
     if (posix_memalign((void *)&bitset->array, sizeof(__m256i),
         ^
[ 20%] Building C object CMakeFiles/roaring.dir/src/containers/containers.c.o
[ 25%] Building C object CMakeFiles/roaring.dir/src/containers/convert.c.o
[ 29%] Building C object CMakeFiles/roaring.dir/src/containers/mixed_intersection.c.o
[ 33%] Building C object CMakeFiles/roaring.dir/src/containers/mixed_union.c.o
[ 37%] Building C object CMakeFiles/roaring.dir/src/containers/run.c.o
[ 41%] Building C object CMakeFiles/roaring.dir/src/roaring.c.o
[ 45%] Building C object CMakeFiles/roaring.dir/src/roaring_array.c.o
[ 50%] Linking C shared library libroaring.so
[ 50%] Built target roaring
Scanning dependencies of target array_container_unit
[ 54%] Building C object CMakeFiles/array_container_unit.dir/tests/array_container_unit.c.o
[ 58%] Linking C executable array_container_unit
[ 58%] Built target array_container_unit
Scanning dependencies of target run_container_unit
[ 62%] Building C object CMakeFiles/run_container_unit.dir/tests/unit.c.o
[ 66%] Linking C executable run_container_unit
[ 66%] Built target run_container_unit
Scanning dependencies of target bitset_container_unit
[ 70%] Building C object CMakeFiles/bitset_container_unit.dir/tests/bitset_container_unit.c.o
[ 75%] Linking C executable bitset_container_unit
[ 75%] Built target bitset_container_unit
Scanning dependencies of target toplevel_unit
[ 79%] Building C object CMakeFiles/toplevel_unit.dir/tests/toplevel_unit.c.o
[ 83%] Linking C executable toplevel_unit
[ 83%] Built target toplevel_unit
Scanning dependencies of target mixed_container_unit
[ 87%] Building C object CMakeFiles/mixed_container_unit.dir/tests/mixed_container_unit.c.o
[ 91%] Linking C executable mixed_container_unit
[ 91%] Built target mixed_container_unit
Scanning dependencies of target unit
[ 95%] Building C object CMakeFiles/unit.dir/tests/unit.c.o
[100%] Linking C executable unit
[100%] Built target unit
Scanning dependencies of target check
UpdateCTestConfiguration  from :/root/CRoaring/DartConfiguration.tcl
UpdateCTestConfiguration  from :/root/CRoaring/DartConfiguration.tcl
Test project /root/CRoaring
Constructing a list of tests
Done constructing a list of tests
Checking test dependency graph...
Checking test dependency graph end
test 1
    Start 1: unit

1: Test command: /root/CRoaring/unit
1: Test timeout computed to be: 9.99988e+06
1: Intel processor:  UNKNOWN	 compiler version: 5.3.0		Build option USEAVX enabled
1: warning, array x array intersection is naive
1: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/census1881
1: Loaded 200 bitmaps from directory benchmarks/realdata/census1881 
1: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/census1881_srt
1: Loaded 200 bitmaps from directory benchmarks/realdata/census1881_srt 
1: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/census-income
1: Loaded 200 bitmaps from directory benchmarks/realdata/census-income 
1: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/census-income_srt
1: Loaded 200 bitmaps from directory benchmarks/realdata/census-income_srt 
1: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/uscensus2000
1: Loaded 200 bitmaps from directory benchmarks/realdata/uscensus2000 
1: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/weather_sept_85
1: Loaded 200 bitmaps from directory benchmarks/realdata/weather_sept_85 
1: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/weather_sept_85_srt
1: Loaded 200 bitmaps from directory benchmarks/realdata/weather_sept_85_srt 
1: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/wikileaks-noquotes
1: Loaded 200 bitmaps from directory benchmarks/realdata/wikileaks-noquotes 
1: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/wikileaks-noquotes_srt
1: Loaded 200 bitmaps from directory benchmarks/realdata/wikileaks-noquotes_srt 
1: [/root/CRoaring/tests/unit.c] your code might be ok.
1/6 Test #1: unit .............................   Passed    4.70 sec
test 2
    Start 2: bitset_container_unit

2: Test command: /root/CRoaring/bitset_container_unit
2: Test timeout computed to be: 9.99988e+06
2: Intel processor:  UNKNOWN	 compiler version: 5.3.0		Build option USEAVX enabled
2: [/root/CRoaring/tests/bitset_container_unit.c] printf_test
2: {1,66,131,202,272}
2: [/root/CRoaring/tests/bitset_container_unit.c] set_get_test
2: [/root/CRoaring/tests/bitset_container_unit.c] and_or_test
2: [/root/CRoaring/tests/bitset_container_unit.c] xor_test
2: [/root/CRoaring/tests/bitset_container_unit.c] andnot_test
2: [/root/CRoaring/tests/bitset_container_unit.c] to_uint32_array_test
2: [/root/CRoaring/tests/bitset_container_unit.c] your code might be ok.
2/6 Test #2: bitset_container_unit ............   Passed    0.01 sec
test 3
    Start 3: array_container_unit

3: Test command: /root/CRoaring/array_container_unit
3: Test timeout computed to be: 9.99988e+06
3: Intel processor:  UNKNOWN	 compiler version: 5.3.0		Build option USEAVX enabled
3: [/root/CRoaring/tests/array_container_unit.c] printf_test
3: {1,2,3,10,10000}
3: [/root/CRoaring/tests/array_container_unit.c] add_contains_test
3: [/root/CRoaring/tests/array_container_unit.c] and_or_test
3: [/root/CRoaring/tests/array_container_unit.c] to_uint32_array_test
3: [/root/CRoaring/tests/array_container_unit.c] your code might be ok.
3/6 Test #3: array_container_unit .............   Passed    0.24 sec
test 4
    Start 4: mixed_container_unit

4: Test command: /root/CRoaring/mixed_container_unit
4: Test timeout computed to be: 9.99988e+06
4: Intel processor:  UNKNOWN	 compiler version: 5.3.0		Build option USEAVX enabled
4: [/root/CRoaring/tests/mixed_container_unit.c] array_bitset_and_or_test
4: [/root/CRoaring/tests/mixed_container_unit.c] your code might be ok.
4/6 Test #4: mixed_container_unit .............   Passed    0.01 sec
test 5
    Start 5: run_container_unit

5: Test command: /root/CRoaring/unit
5: Test timeout computed to be: 9.99988e+06
5: Intel processor:  UNKNOWN	 compiler version: 5.3.0		Build option USEAVX enabled
5: warning, array x array intersection is naive
5: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/census1881
5: Loaded 200 bitmaps from directory benchmarks/realdata/census1881 
5: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/census1881_srt
5: Loaded 200 bitmaps from directory benchmarks/realdata/census1881_srt 
5: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/census-income
5: Loaded 200 bitmaps from directory benchmarks/realdata/census-income 
5: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/census-income_srt
5: Loaded 200 bitmaps from directory benchmarks/realdata/census-income_srt 
5: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/uscensus2000
5: Loaded 200 bitmaps from directory benchmarks/realdata/uscensus2000 
5: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/weather_sept_85
5: Loaded 200 bitmaps from directory benchmarks/realdata/weather_sept_85 
5: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/weather_sept_85_srt
5: Loaded 200 bitmaps from directory benchmarks/realdata/weather_sept_85_srt 
5: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/wikileaks-noquotes
5: Loaded 200 bitmaps from directory benchmarks/realdata/wikileaks-noquotes 
5: [/root/CRoaring/tests/unit.c] realdatacheck benchmarks/realdata/wikileaks-noquotes_srt
5: Loaded 200 bitmaps from directory benchmarks/realdata/wikileaks-noquotes_srt 
5: [/root/CRoaring/tests/unit.c] your code might be ok.
5/6 Test #5: run_container_unit ...............   Passed    4.86 sec
test 6
    Start 6: toplevel_unit

6: Test command: /root/CRoaring/toplevel_unit
6: Test timeout computed to be: 9.99988e+06
6: [/root/CRoaring/tests/toplevel_unit.c] test_printf
6: warning, array x array intersection is naive
6: {1,2,3,100,1000,10000,1000000,20000000}
6: [/root/CRoaring/tests/toplevel_unit.c] test_add
6: [/root/CRoaring/tests/toplevel_unit.c] test_contains
6: [/root/CRoaring/tests/toplevel_unit.c] test_intersection_array_x_array
6: [/root/CRoaring/tests/toplevel_unit.c] test_intersection_array_x_array_inplace
6: [/root/CRoaring/tests/toplevel_unit.c] test_intersection_bitset_x_bitset
6: [/root/CRoaring/tests/toplevel_unit.c] test_intersection_bitset_x_bitset_inplace
6: resultant cardinality is 26666
6: [/root/CRoaring/tests/toplevel_unit.c] test_union
6: [/root/CRoaring/tests/toplevel_unit.c] test_conversion_to_int_array
6: arrays have 97219 and 97219
6:  i=0
6: Container 0 has key 0 and its type is bitset  of card 49999
6:  i=1
6: Container 1 has key 1 and its type is array  of card 3530
6:  i=2
6: Container 2 has key 3 and its type is bitset  of card 43690
6: [/root/CRoaring/tests/toplevel_unit.c] test_array_to_run
6:  i=0
6: Container 0 has key 1 and its type is run  of card 499
6: [/root/CRoaring/tests/toplevel_unit.c] test_array_to_self
6:  i=0
6: Container 0 has key 1 and its type is array  of card 249
6: [/root/CRoaring/tests/toplevel_unit.c] test_bitset_to_run
6:  i=0
6: Container 0 has key 1 and its type is run  of card 49999
6: [/root/CRoaring/tests/toplevel_unit.c] test_bitset_to_self
6:  i=0
6: Container 0 has key 1 and its type is bitset  of card 24999
6: [/root/CRoaring/tests/toplevel_unit.c] test_run_to_self
6:  i=0
6: Container 0 has key 1 and its type is run  of card 49999
6: [/root/CRoaring/tests/toplevel_unit.c] test_remove_run_to_bitset
6:  i=0
6: Container 0 has key 1 and its type is run  of card 49999
6: [/root/CRoaring/tests/toplevel_unit.c] test_remove_run_to_array
6:  i=0
6: Container 0 has key 1 and its type is run  of card 499
6: [/root/CRoaring/tests/toplevel_unit.c] test_conversion_to_int_array_with_runoptimize
6: arrays have 97219 and 97219
6:  i=0
6: Container 0 has key 0 and its type is run  of card 49999
6:  i=1
6: Container 1 has key 1 and its type is array  of card 3530
6:  i=2
6: Container 2 has key 3 and its type is bitset  of card 43690
6: done 17 toplevel tests
6/6 Test #6: toplevel_unit ....................   Passed    0.25 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) =  10.07 sec
[100%] Built target check
```

**Comment by fsaintjacques** [comment]
The problem is with a custom build directory. The problem lies in `set (INCLUDE_FLAGS "-Iinclude -Iinclude/containers -Ibenchmark")`. With cmake you should use the `include_directories`, it will then correctly work.

To reproduce:

```
mkdir build
cd build
cmake ..
make
```

I made it build, but it requires changing some includes.

**Comment by chriso** [comment]
That should be fixed now.

**Comment by lemire** [comment]
@chriso @fsaintjacques 

My unit tests contain some hard-coded path so I expect some kind of failure if you use custom repositories.

https://github.com/RoaringBitmap/CRoaring/blob/master/tests/unit.c#L240

**Comment by lemire** [comment]
(I say "my unit tests" in the sense that I am guilty/to blame, for the unit tests in question.)

**Comment by chriso** [comment]
@lemire 394b9ed5d09a908098509846dfa5f8b9d311c3dd should fix that.

**Comment by lemire** [comment]
@chriso  Indeed.

**Comment by lemire** [comment]
@fsaintjacques 

So, do we merge this?

File: README.md [hunk 2]
```diff
@@ -5,8 +5,16 @@ Bitsets, also called bitmaps, are commonly used as fast data structures. Unfortu
...
-Metamarkets' Druid, Apache Spark, Metamarkets' Druid, Whoosh and eBay's Apache Kylin.
+They are used by several major systems such as [Apache Lucene][lucene] and derivative systems such as [Solr][solr] and
+[Elasticsearch][elasticsearch], [Metamarkets' Druid][druid], [Apache Spark][spark], [Whoosh][whoosh] and eBay's [Apache Kylin][kylin].
+
+[lucene]: https://lucene.apache.org/
+[solr]: https://lucene.apache.org/solr/
+[elasticsearch]: https://www.elastic.co/products/elasticsearch
+[druid]: http://druid.io/
+[spark]: https://spark.apache.org/
+[whoosh]: #
```
Comment (chriso) [inline README.md]: What's Whoosh? I couldn't find a link.

Comment (lemire) [inline README.md]: https://bitbucket.org/mchaput/whoosh/wiki/Home

Comment (lemire) [inline README.md]: https://pypi.python.org/pypi/Whoosh/

Comment (lemire) [inline README.md]: You must not be a Python guy... ;-)

**Comment by lemire** [comment]
@chriso 

I added you to the project and I invited you to join the RoaringBitmap organization.

**Pull request merged**